lambdas/backend-client/src/client.ts

```diff
@@ -0,0 +1,101 @@
+import axios from 'axios';
+import { ITemplateClient } from './types/template-client';
+import {
+  CreateTemplate,
+  Failure,
+  Success,
+  SuccessList,
+  UpdateTemplate,
+} from './types/generated';
+
+type ApiResponse = Failure | Success | SuccessList;
+
+const client = axios.create({
+  baseURL: process.env.TEMPLATE_API_URL,
+});
+
+const failure = (data: Failure) => ({
+  error: {
+    code: data.statusCode,
+    message: data.technicalMessage,
+    details: data.details,
+  },
+});
+
+export const templateClient: ITemplateClient = {
+  createTemplate: async (template: CreateTemplate, token: string) => {
+    const response = await client.post<ApiResponse>('/v1/template', template, {
+      headers: {
+        Authorization: token,
+      },
+    });
+
+    if (response.data.statusCode !== 201) {
+      return failure(response.data as Failure);
+    }
+
+    return {
+      statusCode: response.data.statusCode,
+      data: (response.data as Success).template,
+    };
+  },
+  updateTemplate: async (
+    templateId: string,
+    template: UpdateTemplate,
+    token: string
+  ) => {
+    const response = await client.post<ApiResponse>(
+      `/v1/template/${templateId}`,
+      template,
+      {
+        headers: {
+          Authorization: token,
+        },
+      }
+    );
+
+    if (response.data.statusCode !== 200) {
+      return failure(response.data as Failure);
+    }
+
+    return {
+      statusCode: response.data.statusCode,
+      data: (response.data as Success).template,
+    };
+  },
+  getTemplate: async (templateId: string, token: string) => {
+    const response = await client.get<ApiResponse>(
+      `/v1/template/${templateId}`,
+      {
+        headers: {
+          Authorization: token,
+        },
+      }
+    );
+
+    if (response.data.statusCode !== 200) {
+      return failure(response.data as Failure);
+    }
+
+    return {
+      statusCode: response.data.statusCode,
+      data: (response.data as Success).template,
+    };
+  },
+  listTemplates: async (token: string) => {
+    const response = await client.get<ApiResponse>('/v1/templates', {
+      headers: {
+        Authorization: token,
+      },
+    });
+
+    if (response.data.statusCode !== 200) {
+      return failure(response.data as Failure);
+    }
+
+    return {
+      statusCode: response.data.statusCode,
+      data: (response.data as SuccessList).items,
```

Comment:
the typecast here is unnecessary, you can have client.get<SuccessList>m also in many other places

Reply:
I need this typecasting to access the .items value and similarly the .template property from the API response.

Reply:
the formatting on the previous comment didn't come out right but when you are specifying the type on the client.get line you just need to specify the right type and it will be unnecessary

Reply:
I've updated this. 

We're no longer casting and using the specific type for each api call.